File: src/python/mohair/query/plans.py
#!/usr/bin/env python

# ------------------------------
# License

# Copyright 2023 Aldrin Montana
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


# ------------------------------
# Module Docstring
"""
Classes and functions for interacting with query plans.
"""


# ------------------------------
# Dependencies

# >> Standard libs
from typing import Any

# >> Internal
from mohair import CreateMohairLogger
from mohair.query.types import MohairOp, PipelineOp, BreakerOp
from mohair.query.types import MohairPlan

# ------------------------------
# Module Variables

# >> Logging
logger = CreateMohairLogger(__name__)


# ------------------------------
# Classes

class PlanViewer:
    """ Class for printing or viewing a query plan. """

    @classmethod
    def View(cls, query_plan: MohairPlan, indent='') -> str:
        """ Function to view a plan. """

        logger.debug('Viewing Mohair plan')
        return cls.ViewOp(query_plan.plan_root, indent)

    @classmethod
    def ViewOp(cls, query_op: MohairOp, indent='') -> str:
        """ Function to recursively view operators in a plan. """

        op_prefix = '  '
        if issubclass(type(query_op), BreakerOp): op_prefix = f'\n{indent}'

        input_view_str = cls.ViewInputOps(query_op, indent)
        op_str         = query_op.ViewStr()

        return f'{op_prefix}{op_str}{input_view_str}'

    @classmethod
    def ViewInputOps(cls, query_op: MohairOp, indent='') -> str:
        """ Convenience function to recurse on operator inputs. """

        return ''.join([
            cls.ViewOp(input_op, indent + '  ')
            for input_op in query_op.op_inputs
        ])

    @classmethod
    def ViewBreakers(cls, query_plan: MohairPlan, indent='') -> str:
        """ Function to view the pipeline breakers in a plan. """

        logger.debug('Viewing pipeline breakers')

        # First, the leaves
        # print(type(query_plan.breaker_leaves[0].plan_root))
        # return 'test'
        leaf_breakers = f'\n{indent}'.join([
            str(breaker_op.plan_root)
            for breaker_op in query_plan.breaker_leaves
        ])

        return f'{indent}{leaf_breakers}'


# >> Functions for walking a query plan
class PlanExplorer:

    @classmethod
    def WalkPlan(cls, plan_op: MohairOp) -> MohairPlan:
        """ Function to build a `MohairPlan` pointing to :plan_op:. """

        # Gather a list of ordered `MohairPlan`s for input ops
        input_plans = cls.WalkInputOps(plan_op)
        
        # Compute properties for this `MohairPlan`
        mohair_plan = MohairPlan(plan_op)

        # For everything else we need to compare subplans
        pipeline_len   = 1
        plan_width     = 0
        plan_height    = 0
        breaker_height = 0
        breaker_count  = 0
        breaker_leaves = []  # maybe see if changing these to tuples improves perf.
        breaker_list   = []

        for input_plan in input_plans:
            # plan width (leaf count) is a simple sum
            plan_width  += input_plan.plan_width
            plan_height  = max(plan_height, input_plan.plan_height)

            # breaker count is a simple sum; breaker height is max of breaker heights
            breaker_count  += input_plan.breaker_count
            breaker_height  = max(breaker_height, input_plan.breaker_height)

            # if a subplan has a breaker, it is still a leaf for us
            breaker_leaves.extend(input_plan.breaker_leaves)

            # maintain an iterable list of breakers for easy access
            breaker_list.extend(input_plan.breaker_list)

        # include this operator in properties of the plan
        if issubclass(type(plan_op), BreakerOp):
            breaker_count  += 1
            breaker_height += 1

            # if no other leaves, we are a leaf
            if not breaker_leaves: breaker_leaves.append(mohair_plan)

            # otherwise, we're just a breaker
            else:                  breaker_list.append(mohair_plan)

        # if we had any inputs at all, they're sorted by pipeline length
        if input_plans: pipeline_len = input_plans[0].IncrementPipelineLength()

        # Set plan properties
        mohair_plan.pipeline_len   = pipeline_len
        mohair_plan.plan_width     = plan_width or 1
        mohair_plan.plan_height    = plan_height + 1
        mohair_plan.breaker_count  = breaker_count
        mohair_plan.breaker_height = breaker_height
        mohair_plan.breaker_leaves = breaker_leaves
        mohair_plan.breaker_list   = breaker_list

        return mohair_plan


    @classmethod
    def WalkInputOps(cls, query_op: MohairOp) -> list[MohairPlan]:
        """ Convenience function to recurse on operator inputs. """

        # recurse on input operators
        input_plans = map(cls.WalkPlan, query_op.op_inputs)

        # sort subplans by calling the `IncrementPipelineLength` method
        # this prioritizes extendable pipelines when there are ties in length
        def GetPipeLen(mohair_plan):
            return mohair_plan.IncrementPipelineLength()

        return sorted(input_plans, key=GetPipeLen)

File: src/cpp/mohair/plans.cpp
// ------------------------------
// License
//
// Copyright 2023 Aldrin Montana
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


// ------------------------------
// Dependencies

#include "plans.hpp"


// ------------------------------
// Functions

namespace mohair {

  // >> Debug functions

  template <typename MsgType>
  void PrintProtoMessage(MsgType *msg) {
    string proto_str;

    auto success = TextFormat::PrintToString(*msg, &proto_str);
    if (not success) {
      std::cerr << "Unable to print message" << std::endl;
      return;
    }

    std::cout << "Proto message:" << std::endl
              << proto_str        << std::endl;
  }

  void  PrintSubstraitRel(Rel  *rel_msg ) { PrintProtoMessage<Rel>(rel_msg);   }
  void PrintSubstraitPlan(Plan *plan_msg) { PrintProtoMessage<Plan>(plan_msg); }


  // >> Conversion functions
  unique_ptr<Plan> SubstraitPlanFromString(string &plan_msg) {
    auto substrait_plan = std::make_unique<Plan>();

    substrait_plan->ParseFromString(plan_msg);
    // substrait_plan->PrintDebugString();

    return substrait_plan;
  }

  unique_ptr<Plan> SubstraitPlanFromFile(fstream *plan_fstream) {
    auto substrait_plan = std::make_unique<Plan>();
    if (substrait_plan->ParseFromIstream(plan_fstream)) { return substrait_plan; }

    std::cerr << "Failed to parse substrait plan" << std::endl;
    return nullptr;
  }

  unique_ptr<QueryOp> MohairPlanFrom(Plan *substrait_plan) {
    // walk the top level relations until we find the root (should only be one)
    int root_count = 0;
    unique_ptr<QueryOp> mohair_root;

    for (int ndx = 0; ndx < substrait_plan->relations_size(); ++ndx) {
      PlanRel plan_root { substrait_plan->relations(ndx) };

      // Don't break after we find a RootRel; validate there is only 1
      if (plan_root.rel_type_case() == PlanRel::RelTypeCase::kRoot) {
        ++root_count;
        mohair_root = MohairFrom(plan_root.mutable_root()->mutable_input());
      }
    }

    if (root_count != 1) {
      std::cerr << "Found [" << std::to_string(root_count) << "] RootRels" << std::endl;
      return nullptr;
    }

    return mohair_root;
  }


  // >> Query plan traversal functions

} // namespace: mohair


// ------------------------------
// Classes

namespace mohair {

  // >> Helper functions

  void InsertionSortAppPlans( std::vector<unique_ptr<AppPlan>> &plans
                             ,unique_ptr<AppPlan>               new_plan) {
    auto plans_it = plans.begin();

    // find the spot to insert
    for (; plans_it != plans.end(); plans_it = std::next(plans_it)) {
      if (new_plan->attrs->pipeline_len > (*plans_it)->attrs->pipeline_len) { break; }
    }

    // we always insert, even if its at the end of the list
    plans.insert(plans_it, std::move(new_plan));
  }

  // >> DecomposableProperties functions
  string DecomposableProperties::ToString() {
    std::stringstream prop_stream;

    prop_stream << "Plan properties:"   << std::endl
                << "\tPipeline length:" << pipeline_len   << std::endl
                << "\tPlan width     :" << plan_width     << std::endl
                << "\tPlan height    :" << plan_height    << std::endl
                << "\tBreaker height :" << breaker_height << std::endl
    ;

    return prop_stream.str();
  }

  // PlanVec is a vector of unique_ptr<AppPlan>
  unique_ptr<DecomposableProperties> PropertiesForPlanInputs(PlanVec &plans) {
    auto properties = std::make_unique<DecomposableProperties>();

    // pipeline length, plan width and height, and breaker height
    int p_len = 0, pl_width = 0, pl_height = 0, bheight = 0;

    // local references to vectors
    std::vector<AppPlan*> *break_ops = &(properties->breakers);
    std::vector<AppPlan*> *bleaf_ops = &(properties->breaker_leaves);

    for (size_t ndx = 0; ndx < plans.size(); ++ndx) {
      // >> grab a convenient reference to each plan; these are non-owning
      AppPlan *plan = plans[ndx].get();

      // >> pull in properties from the current plan input
      std::vector<AppPlan*> *plan_breakops = &(plan->attrs->breakers);
      std::vector<AppPlan*> *plan_bleafops = &(plan->attrs->breaker_leaves);

      int plan_plwidth  = plan->attrs->plan_width;
      int plan_plheight = plan->attrs->plan_height;
      int plan_bheight  = plan->attrs->breaker_height;

      int plan_plen = 0;
      if (not plan->plan_op->IsBreaker()) { plan_plen = plan->attrs->pipeline_len; }

      // >> update local variables
      //   |> pointers for fast access
      //      (put this first because it relates to the above check)
      else if (plan->plan_op->IsBreaker()) {
        // A breaker with no leaves must be a leaf itself
        if (plan->attrs->breaker_leaves.size() == 0) {
          bleaf_ops->push_back(plan);
          // bleaf_ops->push_back(std::move(plans[ndx]));
        }

        // Otherwise, it is an internal breaker and we preserve its pointers
        else {
          bleaf_ops->insert(bleaf_ops->cend(), plan_bleafops->cbegin(), plan_bleafops->cend());
          break_ops->insert(break_ops->cend(), plan_breakops->cbegin(), plan_breakops->cend());
          break_ops->push_back(plan);
          // break_ops->push_back(std::move(plans[ndx]));
        }
      }

      //   |> max-type properties
      if (plan_plen     > p_len    ) { p_len     = plan_plen;     }
      if (plan_plheight > pl_height) { pl_height = plan_plheight; }
      if (plan_bheight  > bheight)   { bheight   = plan_bheight;  }

      //   |> sum-type properties
      pl_width += plan_plwidth;
    }

    // >> non-conditional adjustments to properties
    ++p_len;
    ++pl_height;

    // >> conditional adjustments to properties
    if (pl_width == 0) { pl_width = 1; }

    // >> update properties container and return it
    properties->pipeline_len   = p_len;
    properties->plan_height    = pl_height;
    properties->plan_width     = pl_width;
    properties->breaker_height = bheight;

    return properties;
  }

  // >> AppPlan static functions
  std::vector<unique_ptr<AppPlan>> WalkInputOps(QueryOp *parent_op) {
    // Gather child operators
    std::vector<QueryOp *> child_ops = parent_op->GetOpInputs();

    // Initialize a vector of AppPlan instances (wraps QueryOp with tree properties)
    std::vector<unique_ptr<AppPlan>> sorted_plans;
    sorted_plans.reserve(child_ops.size());

    if (child_ops.size() == 0) { return sorted_plans; }

    // Recurse on first child operator; populate first element of sorted_plans
    auto first_childplan = FromPlanOp(child_ops[0]);
    sorted_plans.push_back(std::move(first_childplan));

    // For each input operator to parent_op
    for (size_t child_ndx = 1; child_ndx < child_ops.size(); ++child_ndx) {
      // recurse on child operator
      auto child_plan = FromPlanOp(child_ops[child_ndx]);

      // insertion sort into sorted_plans
      InsertionSortAppPlans(sorted_plans, std::move(child_plan));
    }

    // return AppPlan instances sorted by pipeline length (height-related property)
    return sorted_plans;
  }

  unique_ptr<AppPlan> FromPlanOp(QueryOp *op) {
    // Recurse on input operators
    auto child_plans = WalkInputOps(op);

    /*
    if (   child_plans.size() == 2 && child_plans[0]->plan_op->IsBreaker() && child_plans[0]->plan_op->IsBreaker()) {
      std::cout << "First case of error" << std::endl;
    }
    */

    // Create a plan that wraps this operator
    auto root_plan = std::make_unique<AppPlan>(op);

    // Assign initial properties to root_plan
    root_plan->attrs = PropertiesForPlanInputs(child_plans);

    // Update properties if this operator is a pipeline breaker
    if (op->IsBreaker()) { root_plan->attrs->breaker_height += 1; }

    // Done with traversal
    // NOTE: it seems like this gets called multiple times?
    return root_plan;
  }

  void ViewOp(QueryOp *op, string *indent, std::stringstream *view_stream) {
    // Add indentation for current op
    if (op->IsBreaker()) { (*view_stream) << std::endl << *indent; }
    else                 { (*view_stream)              <<    "  "; }

    // Append representation for this op
    (*view_stream) << op->ViewStr();

    // Increase indent for child ops
    indent->append(2, ' ');

    // Recurse on child ops
    std::vector<QueryOp *> child_ops = op->GetOpInputs();
    for (size_t ndx = 0; ndx < child_ops.size(); ++ndx) {
      ViewOp(child_ops[ndx], indent, view_stream);
    }
  }

  // >> AppPlan member functions
  string AppPlan::ViewPlan() {
    std::stringstream plan_str;
    string            indent { "" };

    ViewOp(plan_op, &indent, &plan_str);

    return plan_str.str();
  }

} // namespace: mohair
